### PR TITLE
adding utils for getting supportted protocols from miner

### PR DESCRIPTION
**Pull request by brossetti1**
utils for working with miner protocols:

- [x] `GetMinerProtocols` - returns list of `MinerProtocols` with distinct multiaddresses for each protocol supportted
- [x] `peerIdFromMultiAddr` - get the peer id from the multiaddress
- [x] `multiaddrToNative` - right now, converts multiaddress to the proper http url for http retrieval
- [x] `minerSupporttedProtocols` - queries the miner to get the supportted protocols (bitswap/http) and their configured addresses

not sure if alot of this is overkill or implements too much logic - my feeling is that `GetMinerProtocols` may have some logic that is unnecessary -- figured you're feedback could be good here

updating go version to 1.19 as go-libp2p requires [atomic types from go packages](https://tip.golang.org/doc/go1.19)

### CODE REVIEW

File: module/indexprovider/utils.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+	gostream "github.com/libp2p/go-libp2p-gostream"
+	"github.com/libp2p/go-libp2p/core/host"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/libp2p/go-libp2p/core/protocol"
+	multistream "github.com/multiformats/go-multistream"
+	"github.com/pkg/errors"
+	"github.com/rs/zerolog"
+)
+
+func QueryRootCid(ctx context.Context, host host.Host, topic string, peerID peer.ID) (cid.Cid, error) {
```
Comment (brossetti1) [inline module/indexprovider/utils.go:21]: extracting this from go-legs so we can remove that dependency

File: module/retrieval/retrieval.go [hunk 2]
```diff
@@ -340,6 +340,8 @@ func (q Auditor) Validate(ctx context.Context, validationInput module.Validation
 					q.log.Info().Str("provider", provider).Str("protocol", string(protocol)).Msg("retrieval succeeded")
 					break
 				}
+			case Bitswap:
```
Comment (brossetti1) [inline module/retrieval/retrieval.go:343]: lint was complaining about this so adding it for now

File: module/retrieval/utils.go [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+	}
+
+	return peerID, nil
+}
+
+// ToURL takes a multiaddr of the form:
+// taken from https://github.com/filecoin-project/go-legs/blame/main/httpsync/multiaddr/convert.go#L43-L84.
+// /dns/thing.com/http/urlescape<path/to/root>.
+// /ip/192.168.0.1/tcp/80/http.
+func ToURL(ma multiaddr.Multiaddr) (*url.URL, error) {
```
Comment (brossetti1) [inline module/retrieval/utils.go:114]: extracting this from go-legs so we can remove that dependency

**Review by xinaxu — APPROVED** [review]

File: module/retrieval/protocolprovider_test.go [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+
+func (t *mockTransportClient) SendQuery(ctx context.Context, p peer.ID) (*types.QueryResponse, error) {
+	return t.protocols, t.err
+}
+
+func TestGetMinerProtocols(t *testing.T) {
+	t.Parallel()
+
+	ma1, err := multiaddr.NewMultiaddr("/ip4/127.0.0.1/tcp/1234/")
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (brossetti1) [reply module/retrieval/protocolprovider_test.go:29]: ok will update this on the bitswap branch

**Pull request merged**